apps/prairielearn/src/pages/studentInstanceQuestionHomework/studentInstanceQuestionHomework.js

```diff
@@ -86,59 +88,110 @@ function processSubmission(req, res, callback) {
+      if (res.locals.assessment.group_work) {
+        const groupConfig = await groupAssessmentHelper.getGroupConfig(res.locals.assessment.id);
+        if (groupConfig.has_roles) {
+          const result = await groupAssessmentHelper.getQuestionPermissions(
+            res.locals.assessment_question.id,
+            res.locals.user.user_id,
+          );
+          // Users without the correct roles cannot submit
+          if (!result.can_submit) {
+            return next(
+              error.make(
+                403,
+                'Current group roles have no permission to submit this question',
+                res.locals,
+              ),
+            );
+          }
+
+          // Get group information
+          const groupId = await groupAssessmentHelper.getGroupId(
+            res.locals.assessment.id,
+            res.locals.user.user_id,
+          );
+          const groupInfo = await groupAssessmentHelper.getGroupInfo(groupId, groupConfig);
+          const { validationErrors, rolesAreBalanced, usersWithoutRoles } = groupInfo.rolesInfo;
+
+          // Prevent submissions if role configuration is invalid
+          if (validationErrors.length > 0 || !rolesAreBalanced || usersWithoutRoles.length > 0) {
+            return next(
+              error.make(
+                403,
+                'Current group role configuration is invalid. Submissions are disabled',
+                res.locals,
+              ),
+            );
+          }
+        }
+      }
+
+      const variant_id = await util.promisify(processSubmission)(req, res);
       res.redirect(
         res.locals.urlPrefix +
           '/instance_question/' +
           res.locals.instance_question.id +
           '/?variant_id=' +
           variant_id,
       );
-    });
-  } else if (req.body.__action === 'attach_file') {
-    util.callbackify(studentInstanceQuestion.processFileUpload)(
-      req,
-      res,
-      function (err, variant_id) {
-        if (ERR(err, next)) return;
-        res.redirect(
-          res.locals.urlPrefix +
-            '/instance_question/' +
-            res.locals.instance_question.id +
-            '/?variant_id=' +
-            variant_id,
-        );
-      },
-    );
-  } else if (req.body.__action === 'attach_text') {
-    util.callbackify(studentInstanceQuestion.processTextUpload)(
-      req,
-      res,
-      function (err, variant_id) {
-        if (ERR(err, next)) return;
-        res.redirect(
-          res.locals.urlPrefix +
-            '/instance_question/' +
-            res.locals.instance_question.id +
-            '/?variant_id=' +
-            variant_id,
-        );
-      },
-    );
-  } else if (req.body.__action === 'delete_file') {
-    util.callbackify(studentInstanceQuestion.processDeleteFile)(
-      req,
-      res,
-      function (err, variant_id) {
+    } else if (req.body.__action === 'attach_file') {
+      util.callbackify(studentInstanceQuestion.processFileUpload)(
```

Comment:
As with the other file, you can directly `await studentInstanceQuestion.processFileUpload` here.